feat(integrations): Add kafka as loader and persist support

wulawulu opened the pull request:
Adds Kafka as a loader and persist support, enabling Swiftide indexing streams to process messages from a Kafka topic to a Kafka topic.

Closes https://github.com/bosun-ai/swiftide/issues/155

timonv commented on swiftide-integrations/src/kafka/mod.rs:
```diff
@@ -0,0 +1,124 @@
+            .context("Failed to create consumer")?;
+        let metadata = consumer.fetch_metadata(Some(&self.topic), None)?;
+        if metadata.topics().is_empty() {
+            let admin_client: AdminClient<DefaultClientContext> = self
```
Afaik in larger setups, often clients don't have admin permission. Maybe we should make this behavior opt in?

timonv commented on swiftide-integrations/src/kafka/persist.rs:
```diff
@@ -0,0 +1,110 @@
+
+    fn batch_size(&self) -> Option<usize> {
+        None
+    }
```
Since you also implemented batch_size, why does this return null? The pipeline will only persist in batch if the batch size is set https://github.com/bosun-ai/swiftide/blob/master/swiftide-indexing/src/pipeline.rs#L344

timonv commented on swiftide-integrations/src/kafka/persist.rs:
```diff
@@ -0,0 +1,110 @@
+                }
+                _ => {
+                    return vec![Err(anyhow::anyhow!(
+                        "persist_value_for_node or persist_key_for_node failed"
```
Could you add some more context to this error? Like the payload/key?

timonv commented on swiftide-integrations/src/kafka/persist.rs:
```diff
@@ -0,0 +1,110 @@
+
+        storage.setup().await.unwrap();
+        storage.store(node.clone()).await.unwrap();
+    }
```
Could you add a test for batch persist?

timonv reviewed (changes requested):
Looking good, thanks for the contribution! Couple questions and comments

wulawulu commented on swiftide-integrations/src/kafka/loader.rs:
```diff
@@ -0,0 +1,181 @@
+
+        consumer
+            .subscribe(&[&topic])
+            .unwrap_or_else(|_| panic!("Failed to subscribe to topic: {topic}"));
```
This return an error,i will fix it.

wulawulu commented on swiftide-integrations/src/kafka/loader.rs:
```diff
@@ -0,0 +1,181 @@
+                            // manually commit offset (if needed)
+                            if let Err(e) = consumer.commit_message(&message, CommitMode::Async) {
+                                tracing::warn!("Failed to commit offset: {:?}", e);
+                            }
```
thank you for the question!

after reviewing this, I realized that this manual commit_message is actually unnecessary in this context. The offset committing strategy should be user-configurable rather than hardcoded here.

I'll go ahead and remove this line and instead rely on the consumer's configuration to handle offset commits.

Thanks again for pointing it out .

wulawulu commented on swiftide-integrations/src/kafka/loader.rs:
```diff
@@ -0,0 +1,181 @@
+
+                            return Some((Ok(node), consumer));
+                        }
+                        tracing::debug!("Skipping message with invalid payload");
```
thank you for the question ,payload are invalid either because the payload was None or because it failed UTF‑8 decoding

wulawulu commented on swiftide-integrations/src/kafka/loader.rs:
```diff
@@ -0,0 +1,181 @@
+
+        let swiftide_stream = stream::unfold(consumer, |consumer| async move {
+            loop {
+                match consumer.recv().await {
```
Your concerns are valid! This design actually handles it well: stream::unfold creates a lazy stream that only calls recv() when downstream requests it, so no blocking occurs. The infinite stream is expected behavior for Kafka, with async await and natural backpressure controlling consumption rate.

wulawulu commented:
Hi @timonv , thank you for the helpful feedback!

I've addressed all your comments in the latest commits.

Please let me know if there's anything else I should improve. Looking forward to your review!

timonv reviewed (changes requested):
Nice, looking good! One small nit on a type, otherwise looks good to go!

wulawulu commented on swiftide-integrations/src/kafka/loader.rs:
```diff
@@ -0,0 +1,175 @@
+            admin
+                .create_topics(
+                    &[NewTopic {
+                        name: topic.into().as_str(),
```
Thanks for the detailed explanation and your patience!
I've updated the code to use `impl AsRef<str>` as suggested to avoid the unnecessary allocation.
